src/expr/src/linear.rs

```diff
@@ -1119,14 +1119,20 @@ impl MapFilterProject {
         let mut reference_count = vec![0; input_arity + self.expressions.len()];
         // Increment reference counts for each use
         for expr in self.expressions.iter() {
-            for col in expr.support().into_iter() {
-                reference_count[col] += 1;
-            }
+            #[allow(deprecated)]
+            expr.visit_post_nolimit(&mut |e| {
+                if let MirScalarExpr::Column(i) = e {
+                    reference_count[*i] += 1;
+                }
+            });
         }
         for (_, pred) in self.predicates.iter() {
-            for col in pred.support().into_iter() {
-                reference_count[col] += 1;
-            }
+            #[allow(deprecated)]
```

Comment:
nit: you can use a worklist approach here because you already defined `fn children()` for `MirScalarExpr` in #20784.

Reply:
Done. I've added a method named `visit_bfs` to abstract the worklist iteration.

Reply:
Why not `visit_pre` to be consistent with the visitor defined on the `Visit` trait? Does name resolution not work as expected in this case?

Reply:
Sorry, I forgot that `visit_pre` is also doing a BFS. I thought it's doing a DFS and calls the visitor preorder. But if the contract of `visit_pre` is just that it will visit a parent node before any of its child nodes, then both approaches satisfy this.

Reply:
Btw. e.g. this website treats BFS in a different category from a preorder DFS:
https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/

Reply:
I'm now thinking about whether the difference between preorder DFS and BFS can affect any optimizer transform, but I can't really come up with any realistic examples where the difference would matter. I guess I'll rename `visit_bfs` to `visit_pre` in my next PR.

Reply:
Yeah, I guess in principle are right. Here is my understanding:

* The recursive `visit_pre` methods in `StackSafeVisit` are DFS pre-order visit.
* The BFS based on a worklist iteration is not really pre-order
* However, both are top-down (parents are visited before children). In the case of BFS this might not happen immediately.

However, the body in the `visit_bfs` method

```rust
let mut worklist = vec![self];
while let Some(e) = worklist.pop() {
    f(e);
    worklist.extend(e.children());
}
```

seems to actually implement pre-order visit because you do `worklist.pop()` in the while loop (so you are removing from the end). For the following three

- 1
  - 2
    - 4
    - 5
  - 3
    - 6
    - 7

you will get this sequence of steps:

| Visit sequence      | `worklist`  |
| ------------------- | ----------- |
|                     | 1           |
| 1                   | 2, 3        |
| 1, 3                | 2, 6, 7     |
| 1, 3, 7             | 2, 6        |
| 1, 3, 7, 6          | 2           |
| 1, 3, 7, 6, 2       | 4, 5        |
| 1, 3, 7, 6, 2, 5    | 4           |
| 1, 3, 7, 6, 2, 5, 4 |             |

I think if you do

```rust
let mut worklist = vec![self];
while let Some(e) = worklist.pop() {
    f(e);
    worklist.extend(e.children().rev());
}
```

you will get order consistent with the `StackSafeVisit::visit_pre` and [the referenced "preorder traversal" example](https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/).

| Visit sequence      | `worklist`  |
| ------------------- | ----------- |
|                     | 1           |
| 1                   | 3, 2        |
| 1, 2                | 3, 5, 4     |
| 1, 2, 4             | 3, 5        |
| 1, 2, 4, 5          | 3           |
| 1, 2, 4, 5, 3       | 7, 6        |
| 1, 2, 4, 5, 3, 6    | 4           |
| 1, 2, 4, 5, 3, 6, 7 |             |

Reply:
Ohh, interesting! You are right, a real BFS would be `pop_front`. With a `pop` from the end and inserting children in reverse order, we really seem to get a DFS preoder. I'll change `visit_bfs` to `visit_pre` and add the `.rev`.